### PR TITLE
Fix and improve reload menu

### PR DESCRIPTION
**Pull request by mugling**
In addition to fixing #15258 and identifying another other bug (58105ac) this provides further enhancements to the reload menu:
- Displayed items are sorted (descending) by remaining ammo
- For empty magazines no statistics are displayed (making full magazines more obvious)

`std::sort` requires both C++11 move semantics and a const `get_item` method so this PR also includes a refactor of `item_location` that:
- Uses C++11 implicit constructors to provide move semantics, including move assignment
- Provides missing const method `item_location::get_item()`
- Makes the class `DefaultConstructible` (replaces `item_location::nowhere`)
- Replaces factory methods with constructor overloads to allow `emplace` and aid templating

### CODE REVIEW

**Comment by mugling** [comment]
I'd like to extend the sorting so that for magazines with equal ammo remaining the nearest is chosen. I don't understand the vehicle code sufficient for this - specifically how do you convert between vehicle part and map position?

File: src/item.cpp [hunk 1]
```diff
@@ -4418,15 +4423,15 @@ item_location item::pick_reload_ammo( player &u, bool interactive ) const
...
-            row += string_format( "| %-7d | %-7d | %-7d | %-7d",
-                                  ammo_damage, ammo_pierce, ammo_range, ammo_dispersion );
+        if( is_gun() || is_magazine() ) {
+            const itype *curammo = it->ammo_data(); // nullptr for empty magazines
+            if( curammo ) {
+                row += string_format( "| %-7d | %-7d | %-7d | %-7d",
+                                      curammo->ammo->damage, curammo->ammo->pierce,
+                                      curammo->ammo->range, curammo->ammo->dispersion );
+            } else {
+                row += _( "|         |         |         |         " );
```
Comment (BevapDin) [inline src/item.cpp]: This row string is translated, the other row string a few lines above isn't. They should either both be translated or both not translated. However, neither of them looks translatable (this one is just formatting), so I suspect it's best to not translate them.

If they get translated, there should be a translator comment for both of them that says the belong together (must have the same width).

**Comment by illi-kun** [comment]
> how do you convert between vehicle part and map position?

Are you asking about ```vehicle::real_global_pos()```, ```vehicle::global_part_pos3( const int &index )```, etc. or you need something else?

**Comment by mugling** [comment]
I need to convert a vehicle plus part number into map position, so `vehicle::global_part_pos3` looks like the right item

**Comment by mugling** [comment]
This and #15256 are merge conflicts and I need the functionality from both to implement `std::vector<item_location> visitable<T>::find_ammo( item& gun )`

File: src/item.cpp [hunk 2]
```diff
@@ -4418,15 +4423,15 @@ item_location item::pick_reload_ammo( player &u, bool interactive ) const
...
-            const auto ammo_dispersion = curammo ? 100 - curammo->ammo->dispersion : 0;
-
-            row += string_format( "| %-7d | %-7d | %-7d | %-7d",
-                                  ammo_damage, ammo_pierce, ammo_range, ammo_dispersion );
+        if( is_gun() || is_magazine() ) {
+            const itype *curammo = it->ammo_data(); // nullptr for empty magazines
+            if( curammo ) {
+                row += string_format( "| %-7d | %-7d | %-7d | %-7d",
+                                      curammo->ammo->damage, curammo->ammo->pierce,
+                                      curammo->ammo->range, curammo->ammo->dispersion );
```
Comment (BevapDin) [inline src/item.cpp]: This now shows `dispersion`, previously it showed `100 - dispersion`, I think the previous version is the correct one as the row is titled "accuracy" (through I'm not sure if that's the inverse of dispersion).

Comment (mugling) [inline src/item.cpp]: Well noticed, fixed in #7bd2c01

**Comment by BevapDin** [comment]
> Displayed items are sorted (descending) by remaining ammo

Except for quivers with ammo in it. Their content is also listed twice: once as simple ammo item and once as "quiver with ammo". Sorting isn't done for those at all as `ammo_remaining` returns 0 for ammo items itself and for quivers. This is not a stopper.

But this one is:

Your changes in game.cpp change the behavior of eating food in containers. Set the "Drop empty containers" option to All and drink out a water bottle or can or whatever. The games says "You leave the empty [container]" and leaves the container in the inventory. It's supposed to drop it. This would be done in `player::consume(int)`, but you removed the call to that function and now `player::consume_item(item*)` is called, which does not handle the dropping of containers.

**Comment by mugling** [comment]
> Except for quivers with ammo in it

Indeed, see #15271. I know the source of that but I want to wait for #15256 and overhaul the entire ammo search algorithm as it's got a few outstanding issues open against it.

**Comment by mugling** [comment]
> Your changes in game.cpp change the behavior of eating food in containers

Yes, appears to be regression, will look into that

**Pull request merged**